### DISCUSSION TITLE
Adding, building and linking Lua in a project

### DISCUSSION THREAD

**Discussion by oscarbaselga** [zephyrproject-rtos/zephyr#45485] [Q&A] [answered]
**Bug description**
My project has to integrate the Lua scripting language to implement user services. For that purpose, I first included its source code to the project, then coded cmake to build this code and to link the library to the final target, and finally invoked the C API of Lua from the main function. This does not work, it failed at compile time and run time.

My environment:
- NXP FRDM K64F
- Zephyr v2.7.0
- West v0.12.0
- Zephyr SDK v0.13.1
- Cmake v3.22.3
- Linux host system

**To reproduce the problem**
1. Project tree
```
tfm
├── CMakeLists.txt
├── lib
│   └── lua-5.4.4
│       ├── lapi.c
│       ├── lapi.h
│       ├── lauxlib.c
│       ├── lauxlib.h
│       ├── lbaselib.c
│       ├── lcode.c
│       ├── lcode.h
│       ├── lcorolib.c
│       ├── lctype.c
│       ├── lctype.h
│       ├── ldblib.c
│       ├── ldebug.c
│       ├── ldebug.h
│       ├── ldo.c
│       ├── ldo.h
│       ├── ldump.c
│       ├── lfunc.c
│       ├── lfunc.h
│       ├── lgc.c
│       ├── lgc.h
│       ├── linit.c
│       ├── liolib.c
│       ├── ljumptab.h
│       ├── llex.c
│       ├── llex.h
│       ├── llimits.h
│       ├── lmathlib.c
│       ├── lmem.c
│       ├── lmem.h
│       ├── loadlib.c
│       ├── lobject.c
│       ├── lobject.h
│       ├── lopcodes.c
│       ├── lopcodes.h
│       ├── lopnames.h
│       ├── loslib.c
│       ├── lparser.c
│       ├── lparser.h
│       ├── lprefix.h
│       ├── lstate.c
│       ├── lstate.h
│       ├── lstring.c
│       ├── lstring.h
│       ├── lstrlib.c
│       ├── ltable.c
│       ├── ltable.h
│       ├── ltablib.c
│       ├── ltm.c
│       ├── ltm.h
│       ├── luaconf.h
│       ├── lua.h
│       ├── lua.hpp
│       ├── lualib.h
│       ├── lundump.c
│       ├── lundump.h
│       ├── lutf8lib.c
│       ├── lvm.c
│       ├── lvm.h
│       ├── lzio.c
│       └── lzio.h
├── prj.conf
├── README.rst
├── sample.yaml
└── src
    └── main.c
```
3. Project code
3.1. `main.c`
```
#include <zephyr.h>
#include <fs/fs.h>

#include <stdlib.h>

#include "lua.h"
#include "lauxlib.h"
#include "lualib.h"

void main(void)
{
	lua_State *L = luaL_newstate();
  	luaL_openlibs(L);
  	luaL_dostring(L, "print('Hello from Lua')");

	printk("Hello with Lua!\n");

	lua_close(L);
}
```
2.2. `prj.conf`
```
CONFIG_NEWLIB_LIBC=y
CONFIG_FILE_SYSTEM=y
CONFIG_MAIN_STACK_SIZE=8192
```
2.3. `CMakeLists.txt`
```
cmake_minimum_required(VERSION 3.20.0)
find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE})
project(tfm)

# Application source
target_sources(app PRIVATE src/main.c)

file(GLOB lua_sources lib/lua-5.4.4/*.c)
file(GLOB lua_headers lib/lua-5.4.4/*.h)

target_sources(app PRIVATE ${lua_sources} ${lua_headers})

target_compile_options(app PRIVATE -std=gnu99
    -O2 -Wall -Wextra -DLUA_COMPAT_5_3)
target_link_options(app PUBLIC -lm)
target_include_directories(app PRIVATE lib/lua-5.4.4)
```
3. Compilation command: `west build -b frdm_k64f sample/tfm`

**Expected behavior**
I expect the project compiles and, when flashing and monitoring, the following output is displayed:
```
Hello from Lua
Hello with Lua!
```

**Impact**
It is absolutely necessary that this works, otherwise my project is useless.

**Logs and console output**
When building the project, the error shown is:
```
[3/9] Linking C executable zephyr/zephyr_prebuilt.elf
FAILED: zephyr/zephyr_prebuilt.elf zephyr/zephyr_prebuilt.map /home/oscar/zephyr_home/zephyr/build/zephyr/zephyr_prebuilt.map 
: && ccache /home/oscar/zephyr-sdk-0.13.1/arm-zephyr-eabi/bin/arm-zephyr-eabi-gcc  -lm zephyr/CMakeFiles/zephyr_prebuilt.dir/misc/empty_file.c.obj -o zephyr/zephyr_prebuilt.elf  zephyr/CMakeFiles/offsets.dir/./arch/arm/core/offsets/offsets.c.obj  -fuse-ld=bfd  -Wl,-T  zephyr/linker_zephyr_prebuilt.cmd  -Wl,-Map=/home/oscar/zephyr_home/zephyr/build/zephyr/zephyr_prebuilt.map  -Wl,--whole-archive  app/libapp.a  zephyr/libzephyr.a  zephyr/arch/common/libarch__common.a  zephyr/arch/arch/arm/core/aarch32/libarch__arm__core__aarch32.a  zephyr/arch/arch/arm/core/aarch32/cortex_m/libarch__arm__core__aarch32__cortex_m.a  zephyr/arch/arch/arm/core/aarch32/mpu/libarch__arm__core__aarch32__mpu.a  zephyr/lib/libc/newlib/liblib__libc__newlib.a  zephyr/lib/posix/liblib__posix.a  zephyr/boards/arm/frdm_k64f/libboards__arm__frdm_k64f.a  zephyr/subsys/fs/libsubsys__fs.a  zephyr/drivers/clock_control/libdrivers__clock_control.a  zephyr/drivers/console/libdrivers__console.a  zephyr/drivers/gpio/libdrivers__gpio.a  zephyr/drivers/serial/libdrivers__serial.a  zephyr/drivers/timer/libdrivers__timer.a  modules/hal_nxp/hal_nxp/lib..__modules__hal__nxp.a  -Wl,--no-whole-archive  zephyr/kernel/libkernel.a  -L"/home/oscar/zephyr-sdk-0.13.1/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/10.3.0/thumb/v7e-m/nofp"  -L/home/oscar/zephyr_home/zephyr/build/zephyr  -lgcc  zephyr/arch/common/libisr_tables.a  -mcpu=cortex-m4  -mthumb  -mabi=aapcs  -mfp16-format=ieee  -Wl,--gc-sections  -Wl,--build-id=none  -Wl,--sort-common=descending  -Wl,--sort-section=alignment  -Wl,-u,_OffsetAbsSyms  -Wl,-u,_ConfigAbsSyms  -nostdlib  -static  -no-pie  -Wl,-X  -Wl,-N  -Wl,--orphan-handling=warn  -lm  -Wl,-lc  -L"/home/oscar/zephyr-sdk-0.13.1/arm-zephyr-eabi/arm-zephyr-eabi"/lib/thumb/v7e-m/nofp  -Wl,-lgcc  -lc  -specs=nano.specs && cd /home/oscar/zephyr_home/zephyr/build/zephyr && /opt/cmake-3.22.3-linux-x86_64/bin/cmake -E echo
/home/oscar/zephyr-sdk-0.13.1/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/10.3.0/../../../../arm-zephyr-eabi/bin/ld.bfd: /home/oscar/zephyr-sdk-0.13.1/arm-zephyr-eabi/arm-zephyr-eabi/lib/thumb/v7e-m/nofp/libc_nano.a(lib_a-unlinkr.o): in function `_unlink_r':
unlinkr.c:(.text._unlink_r+0xc): undefined reference to `_unlink'
/home/oscar/zephyr-sdk-0.13.1/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/10.3.0/../../../../arm-zephyr-eabi/bin/ld.bfd: /home/oscar/zephyr-sdk-0.13.1/arm-zephyr-eabi/arm-zephyr-eabi/lib/thumb/v7e-m/nofp/libc_nano.a(lib_a-linkr.o): in function `_link_r':
linkr.c:(.text._link_r+0xe): undefined reference to `_link'
/home/oscar/zephyr-sdk-0.13.1/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/10.3.0/../../../../arm-zephyr-eabi/bin/ld.bfd: /home/oscar/zephyr-sdk-0.13.1/arm-zephyr-eabi/arm-zephyr-eabi/lib/thumb/v7e-m/nofp/libc_nano.a(lib_a-timesr.o): in function `_times_r':
timesr.c:(.text._times_r+0x2): undefined reference to `_times'
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.
FATAL ERROR: command exited with status 1: /usr/local/bin/cmake --build /home/oscar/zephyr_home/zephyr/build
```
It seems that zephyr is not able to find the functions `_unlink`, `_link` and `_times`, so if I add a simple implementation in the `main.c` it works:
```
clock_t _times(struct tms* tms)
{
	errno = ENOSYS;
	return (clock_t) -1;
}

int _unlink(const char *pathname)
{
	return fs_unlink(pathname);
}

int _link(const char *oldpath, const char *newpath)
{
	errno = ENOSYS;
	return -1;
}
```
However, once the binary is flashed, a run time error appears:
```
[00:00:00.009,000] <err> os: ***** BUS FAULT *****
[00:00:00.009,000] <err> os:   Stacking error
[00:00:00.009,000] <err> os:   Imprecise data bus error
[00:00:00.009,000] <err> os:   NXP MPU error, port 3
[00:00:00.009,000] <err> os:     Mode: Supervisor, Data Address: 0x20000a08
[00:00:00.009,000] <err> os:     Type: Write, Master: 0, Regions: 0x8200
[00:00:00.009,000] <err> os: r0/a1:  0xaaaaaaaa  r1/a2:  0xaaaaaaaa  r2/a3:  0x00027556
[00:00:00.009,000] <err> os: r3/a4:  0x01000000 r12/ip:  0x20000084 r14/lr:  0x20000088
[00:00:00.009,000] <err> os:  xpsr:  0x0001b800
[00:00:00.009,000] <err> os: Faulting instruction address (r15/pc): 0x200004c8
[00:00:00.009,000] <err> os: >>> ZEPHYR FATAL ERROR 2: Stack overflow on CPU 0
[00:00:00.009,000] <err> os: Current thread: 0x20000490 (unknown)
[00:00:00.091,000] <err> os: Halting system
```

**Additional context**
In this repository: [https://github.com/Mohren/zephyrlua/tree/master/hello_world_lua5_3_5](https://github.com/Mohren/zephyrlua/tree/master/hello_world_lua5_3_5), the author says that the compilation error is caused by the `os` and `io` Lua libs, so he remove them, but this is not a good practice.

**Comment by stephanosio** [comment]
This is not a Zephyr bug, but a problem/bug with something you are working on.

**Comment by cfriedt** [comment]
Funny that this is the top result when googling for Lua support in Zephyr. It should be done, and I think it can be done relatively easily.

**Comment by tenllado** [comment, marked as answer]
I could make this example run with the following project layout:

CMakeLists.txt
prj.conf
src
lua

The content of the CMakeLists.txt file is

cmake_minimum_required(VERSION 3.20.0)

find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE})
project(sol)

add_library(liblua STATIC
	lua/lapi.c
	lua/lauxlib.c
	lua/lbaselib.c
	lua/lcode.c
	lua/lcorolib.c
	lua/lctype.c
	lua/ldblib.c
	lua/ldebug.c
	lua/ldo.c
	lua/ldump.c
	lua/lfunc.c
	lua/lgc.c
	lua/linit.c
	lua/liolib.c
	lua/llex.c
	lua/lmathlib.c
	lua/lmem.c
    lua/loadlib.c
	lua/lobject.c
	lua/lopcodes.c
    lua/loslib.c
	lua/lparser.c
	lua/lstate.c
	lua/lstring.c
	lua/lstrlib.c
	lua/ltable.c
	lua/ltablib.c
	lua/ltm.c
	lua/lundump.c
    lua/lutf8lib.c
	lua/lvm.c
	lua/lzio.c
	)

target_include_directories(liblua
	INTERFACE ${CMAKE_CURRENT_SOURCE_DIR}/lua
	PRIVATE ${CMAKE_CURRENT_SOURCE_DIR}/lua
	)

target_link_libraries(liblua zephyr_interface)

install(TARGETS liblua DESTINATION lib)

add_subdirectory(src)

The content of the prj.conf is:

CONFIG_APPLICATION_DEFINED_SYSCALL=y
CONFIG_PRINTK=y
CONFIG_ASSERT=y
CONFIG_NEWLIB_LIBC=y
CONFIG_NEWLIB_LIBC_FLOAT_PRINTF=y
CONFIG_NEWLIB_LIBC_FLOAT_SCANF=y
CONFIG_INIT_STACKS=y
CONFIG_MAIN_STACK_SIZE=4096
CONFIG_LOG=y
CONFIG_LOG_MODE_MINIMAL=y
CONFIG_FILE_SYSTEM=y

In the src file I have:
CMakeLists.txt
main.c
missing_stubs.c

where CMakeLists.txt contains:

target_sources(app
        PRIVATE main.c missing_stubs.c
        )

target_link_libraries(app
        PRIVATE liblua
        )

main.c contains:

#include <zephyr.h>
#include <fs/fs.h>

#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include "lua.h"
#include "lauxlib.h"
#include "lualib.h"

void main(void)
{
        lua_State *L;

        printk("Starting lua code\n");
        L = luaL_newstate();
        luaL_openlibs(L);
        luaL_dostring(L, "print('Hello from Lua')");
        lua_close(L);
        printk("Test Done!\n");
}

and missing_stubs.c contains:

#include <zephyr.h>
#include <stdio.h>

/* FIXME: this stubs should be properly implemented, in zephyr itself or here.
 * This is just a temporal non-working implementation to start testing lua with
 * zephyr.
 */

clock_t _times(struct tms* tms)
{
        errno = ENOSYS;
        return (clock_t) -1;
}

int _unlink(const char *pathname)
{
        return fs_unlink(pathname);
}


int _link(const char *oldpath, const char *newpath)
{
        errno = ENOSYS;
        return -1;
}

The code runs on an nxp discovery l475 iot1 board, it shows the lua string:

*** Booting Zephyr OS build 70cca36acc29  ***
Starting lua code
Hello from Lua
Test Done!

The content of the lua directory is the sources of lua downloaded from their git repository.

I could not solve however the problem of the missing newlib stubs. I think these should be added to the newlib library from zephyr.

Hope this helps.

Christian Tenllado

**Comment by tenllado** [comment]
For your convenience I created a github repo with the code for the lua hello world example in a west workspace. The repo is https://github.com/tenllado/zephyr_hello_world_lua.git

The README.md file includes the instructions to compile and run it. I will appreciate any help with the missing newlib stubs problem.

**Comment by cfriedt** [comment]
@oscarbaselga - awesome!

It would probably be worth it to consider adding a `cmake/` directory and something like `lua.cmake` there, so that `.lua` files from anywhere could be concatenated together in `build/generated/all_lua_sources.lua` or something, then compiled into bytecode with `luac`, converted to a C array with `xxd` (or similar), linked-in to C via `#include`, and then e.g. loaded from memory inside of a `SYS_INIT()` call with `luaL_loadstring()`. There is the `CONFIG_POSIX_FS` option as well, so `.luac` bytecode or `.lua` source could be loaded from a file as well.

E.g. application CMakeLists.txt
```cmake
# concatenate feature.lua to build/generated/all_lua_sources.lua
lua_sources_ifdef(CONFIG_SOME_LUA_FEATURE feature.lua)

# Export one or more function definitions so that they can be added to a C header prior to compiling C code
# Maybe even automatically generate C wrappers for Lua functions?
# e.g. z = f(x, y) => double zlua_f(int x, double y)
# https://www.lua.org/pil/25.2.html
lua_export_ifdef(CONFIG_SOME_LUA_FEATURE "double", "feature_function", "int", "x", "double", "y")
```

C / C++ file
```cpp
#include <generated/all_lua_functions.h>

void arbitrary_function() {
  int x = 42;
  double y = 73.0;
  double z = zlua_f(x, y);
}
```

I guess CMake could be used to collect all of the "entry points" into lua code (so that it isn't just a `main()` function). That way, a single header can be included for C to have prototypes for `Lual_<function>` calls.

I would put the lua library definition inside of `lib/lua/CMakeLists.txt`.

What you've done with the kind of "west submodule" for the Lua source code is smart - it makes your Lua module really a thin wrapper around upstream. If `west` supported git submodules that would probably make things easier.

**Comment by rodrigopex** [comment]
@oscarbaselga @tenllado, I was wondering why you did not follow up with the Lua module on Zephyr. I am investing some time in that and would like to understand why such a cool feature is still not there, given that you had a good base code 3 years ago. 

Are you still interested in collaborating with this effort?

**Reply by tenllado** [reply]
@rodrigopex we are interested, we indeed have a lua module for zephyr that we are using in our project, it is working very well. I think we can export it to a public github repository if there is interest in the zephyr community for it.

**Reply by rodrigopex** [reply]
@tenllado 

> @rodrigopex we are interested, we indeed have a lua module for zephyr that we are using in our project, it is working very well. I think we can export it to a public github repository if there is interest in the zephyr community for it.

Very good. We have an initial effort mapped in this repository https://github.com/rodrigopex/lua_zephyr

Let's see the diffs and get the better approach from both and submit. What do you think? Maybe we can schedule a meeting for that.

**Reply by tenllado** [reply]
@rodrigopex 

A meeting with you would be great. We have some close deadlines from our project that we have to work on before summer vacations. After that we will export the lua module to a public repository and document its usage. We can schedule a meeting with you after that if you agree.
